sound/soc/codecs/rt715.c

```diff
@@ -222,9 +222,238 @@ static int rt715_set_amp_gain_get(struct snd_kcontrol *kcontrol,
+		RT715_SET_GAIN_LINE_ADC_L, RT715_SET_GAIN_MIX_ADC_L,
+		RT715_SET_GAIN_MIX_ADC2_L};
+	unsigned int addr_h, addr_l, val_h = 0x0, val_ll, val_lr;
+	unsigned int read_ll, read_rl, i, j, loop_cnt = 4;
+	unsigned int k_shift = RT715_DIR_IN_SFT, k_max = 1;
+
+	for (j = 0; j < loop_cnt; j++) {
+		/* Can't use update bit function, so read the original value first */
+		addr_h = capture_reg_H[j];
+		addr_l = capture_reg_L[j];
+		rt715_get_gain(rt715, addr_h, addr_l, val_h, &read_rl, &read_ll);
+
+		if (dapm->bias_level <= SND_SOC_BIAS_STANDBY)
+			regmap_write(rt715->regmap,
+					RT715_SET_AUDIO_POWER_STATE, AC_PWRST_D0);
+
+		/* L Channel */
+		/* for mute */
+		val_ll = (k_max - ucontrol->value.integer.value[j * 2]) << 7;
+		/* keep gain */
+		read_ll = read_ll & 0x7f;
+		val_ll |= read_ll;
+
+		/* R Channel */
+		/* for mute */
+		val_lr = (k_max - ucontrol->value.integer.value[j * 2 + 1]) << 7;
+		/* keep gain */
+		read_rl = read_rl & 0x7f;
+		val_lr |= read_rl;
+
+		for (i = 0; i < 3; i++) { /* retry 3 times at most */
+
+			if (val_ll == val_lr) {
+				/* Set both L/R channels at the same time */
+				val_h = (1 << k_shift) | (3 << 4);
+				regmap_write(rt715->regmap, addr_h,
+					(val_h << 8 | val_ll));
+				regmap_write(rt715->regmap, addr_l,
+					(val_h << 8 | val_ll));
+			} else {
+				/* Lch*/
+				val_h = (1 << k_shift) | (1 << 5);
+				regmap_write(rt715->regmap, addr_h,
+					(val_h << 8 | val_ll));
+				/* Rch */
+				val_h = (1 << k_shift) | (1 << 4);
+				regmap_write(rt715->regmap, addr_l,
+					(val_h << 8 | val_lr));
+			}
+			val_h = 0x0;
+			rt715_get_gain(rt715, addr_h, addr_l, val_h,
+					&read_rl, &read_ll);
+			if (read_rl == val_lr && read_ll == val_ll)
+				break;
+		}
+	}
+	/* D0:power on state, D3: power saving mode */
+	if (dapm->bias_level <= SND_SOC_BIAS_STANDBY)
+		regmap_write(rt715->regmap,
+				RT715_SET_AUDIO_POWER_STATE, AC_PWRST_D3);
+	return 0;
+}
+
+static int rt715_set_main_switch_get(struct snd_kcontrol *kcontrol,
+				  struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int capture_reg_H[] = {RT715_SET_GAIN_MIC_ADC_H,
+		RT715_SET_GAIN_LINE_ADC_H, RT715_SET_GAIN_MIX_ADC_H,
+		RT715_SET_GAIN_MIX_ADC2_H};
+	unsigned int capture_reg_L[] = {RT715_SET_GAIN_MIC_ADC_L,
+		RT715_SET_GAIN_LINE_ADC_L, RT715_SET_GAIN_MIX_ADC_L,
+		RT715_SET_GAIN_MIX_ADC2_L};
+	unsigned int addr_h, addr_l, val_h = 0x0, i, loop_cnt = 4;
+	unsigned int read_ll[4] = {0}, read_rl[4] = {0};
+
+	for (i = 0; i < loop_cnt; i++) {
+		addr_h = capture_reg_H[i];
+		addr_l = capture_reg_L[i];
+		rt715_get_gain(rt715, addr_h, addr_l, val_h, &read_rl[i], &read_ll[i]);
+
+		/* for mute status */
+		read_ll[i] = !(read_ll[i] & 0x80);
+		read_rl[i] = !(read_rl[i] & 0x80);
+
+		ucontrol->value.integer.value[i * 2] = read_ll[i];
+		ucontrol->value.integer.value[i * 2 + 1] = read_rl[i];
```

Comment:
how about
```
		ucontrol->value.integer.value[i * 2] = !(read_ll & 0x80);
		ucontrol->value.integer.value[i * 2 + 1] = !(read_rl & 0x80);
```
BTW, I had a look at `rt715_get_gain()` and it isn't very safe:
```
static void rt715_get_gain(struct rt715_priv *rt715, unsigned int addr_h,
				unsigned int addr_l, unsigned int val_h,
				unsigned int *r_val, unsigned int *l_val)
{
	int ret;
	/* R Channel */
	*r_val = (val_h << 8);
```
Why is this done? This isn't needed for the follow-up `reg_read()` call. Not to mention the superfluous parentheses ;-) This means, that the output value is *changed* even if the below `reg_read()` fails:
```
	ret = regmap_read(rt715->regmap, addr_l, r_val);
	if (ret < 0)
		pr_err("Failed to get R channel gain.\n");
```
Is this how it's supposed to be? So was it supposed to be
```
	ret = regmap_read(rt715->regmap, addr_l, r_val);
	if (ret < 0) {
		pr_err("Failed to get R channel gain.\n");
		*r_val = val_h << 8;
	}
```
? Same for the left channel

Reply:
Because we have our own reg_read callback in regmap_config, so operation above are needed. Please refer "static const struct regmap_config rt715_regmap" in rt715-sdw.c. Thanks.

Reply:
oh... ok, but it's still not a good idea to modify input-output arguments when the function is failing. So I'd modify `rt715_get_gain()` to something like
```
static int rt715_get_gain(struct rt715_priv *rt715, unsigned int addr_h,
				unsigned int addr_l, unsigned int val_h,
				unsigned int *r_val, unsigned int *l_val)
{
	int ret;
	unsigned int tmp_r, tmp_l;

	tmp_r = val_h << 8;
	tmp_l = tmp_r | 0x2000;

	/* R Channel */
	ret = regmap_read(rt715->regmap, addr_l, &tmp_r);
	if (ret < 0) {
		pr_err("Failed to get R channel gain.\n");
		return ret;
	}

	/* L Channel */
	ret = regmap_read(rt715->regmap, addr_h, &tmp_l);
	if (ret < 0) {
		pr_err("Failed to get L channel gain.\n");
		return ret;
	}

	*r_val = tmp_r;
	*l_val = tmp_l;

	return 0;
}
```
And then check for return value where this function is called. Could you cook up a separate fixup PR for this?

Reply:
Okay, I'll send seperate PR for it after this PR merged.